rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,132 @@
+CREATE INDEX events_search_session_events_idx ON events (session_id, event_time, event_id) WHERE session_id != '00000000-0000-0000-0000-000000000000';
+```
+
+While the database schema is compatible with the one for the backend, we are going to strongly discourage running the backend and the audit log in the same database (using two different databases in the same instance of Postgres is ok), because logical decoding is per-database, not per-table, and keeping both in the same database will require extra work to filter out the changes to the audit log - especially when sifting through the transactions deleting events that are past the retention period.
+
+The `event_id` column is a UUID chosen at random at insert time, unrelated to the event data, only used to disambiguate events for the purpose of pagination; `creation_time`, likewise, is only used for deleting events after some retention period (defaulting to one year).
+
+With minor modifications, this schema can support range based partitioning by `event_time`, creating partitions by hand or using `pg_partman`. The initial implementation will not use partitioning, but setting up the table layout for it will allow for a much easier migration if and when we end up supporting it.
+
+### Requirements and recommendations
+
+Any supported version of Postgres (11 to 15) should work, with `wal2json` 2.1 or later, although for better results, Postgres 13 or higher is recommended (as logical decoding had some internal improvements).
+
+Because the db user that Teleport connects as needs `REPLICATION` permissions, which effectively grants read access to the entire database cluster, the backend should not be used in a database cluster that also contains data that the Teleport auth server should not have access to (it's fine to use a single cluster for the backend and the audit log storage, for instance).
+
+Because of how temporary replication slots (and prepared statements) are tied to individual connections, the use of external connection poolers (such as `pgbouncer`) is not supported.
+
+For self-managed Postgres setups we recommend the use of [certificate authentication](https://www.postgresql.org/docs/current/auth-cert.html); in Azure we support and recommend the use of Azure AD authentication over the use of hardcoded passwords.
+
+Teleport will attempt to create the databases specified in the configuration for itself, and set them up with the correct schema (as well as apply any future migrations, if needed). It's possible to create the databases manually and not give Teleport permissions to create databases, and technically it's even possible to manually set up the database schemas (to avoid granting Teleport the ability to create tables, for instance), but we recommend giving Teleport its own database for the backend - and since Teleport will need read-write access to all the data in it anyway, we also recommend letting Teleport manage the schema automatically.
```

Comment:
If possible, we should require teleport be allowed to manage the schema... or at least not document the fact that a user-provided schema would work.  Seems like it'd be a real pain to maintain compatibility with externally managed schemas if we ever needed to migrate/fix something about the schema.

Reply:
The issue with that is that if Teleport has the ability to apply schema changes, it almost surely needs the ability to delete or update the events in the audit log - maybe we can require owner on the backend database and allow the user to restrict access to the events database?